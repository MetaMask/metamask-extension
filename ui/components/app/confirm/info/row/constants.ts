export const TEST_ADDRESS = '0x5CfE73b6021E818B776b421B1c4Db2474086a7e1';

export enum RowAlertKey {
  EstimatedFee = 'estimatedFee',
  SigningInWith = 'signingInWith',
  RequestFrom = 'requestFrom',
  Resimulation = 'resimulation',
  Speed = 'speed',
}

export enum AlertActionKey {
  Buy = 'buy',
  ShowAdvancedGasFeeModal = 'showAdvancedGasModal',
  ShowGasFeeModal = 'showGasFeeModal',
}
