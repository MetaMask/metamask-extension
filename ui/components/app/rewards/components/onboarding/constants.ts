export const REWARDS_ONBOARD_OPTIN_LEGAL_LEARN_MORE_URL =
  'https://support.metamask.io/manage-crypto/metamask-rewards';

export const REWARDS_ONBOARD_TERMS_URL = 'https://go.metamask.io/rewards-terms';
