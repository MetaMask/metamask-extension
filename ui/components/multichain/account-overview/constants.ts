export const FUND_SLIDE = {
  id: 'fund',
  title: 'slideFundWalletTitle',
  description: 'slideFundWalletDescription',
  image: './images/slide-fund-icon.svg',
  href: 'https://portfolio.metamask.io/buy/build-quote',
};

///: BEGIN:ONLY_INCLUDE_IF(build-main,build-beta,build-flask)
export const BRIDGE_SLIDE = {
  id: 'bridge',
  title: 'slideBridgeTitle',
  description: 'slideBridgeDescription',
  image: './images/slide-bridge-icon.svg',
};
///: END:ONLY_INCLUDE_IF

export const CARD_SLIDE = {
  id: 'card',
  title: 'slideDebitCardTitle',
  description: 'slideDebitCardDescription',
  image: './images/slide-card-icon.svg',
  href: 'https://portfolio.metamask.io/card',
};

export const CASH_SLIDE = {
  id: 'cash',
  title: 'slideCashOutTitle',
  description: 'slideCashOutDescription',
  image: './images/slide-sell-icon.svg',
  href: 'https://portfolio.metamask.io/sell',
};

export const ZERO_BALANCE = '0x00';
