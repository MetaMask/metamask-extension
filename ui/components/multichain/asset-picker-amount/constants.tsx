export const LARGE_SYMBOL_LENGTH = 7;
