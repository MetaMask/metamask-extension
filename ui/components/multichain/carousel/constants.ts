export const MARGIN_VALUES = {
  SLIDE_BOTTOM: '40px',
  CONTAINER_SIDE: '16px',
  SLIDE_GAP: '2%',
  ZERO: '0',
};

export const WIDTH_VALUES = {
  FULL_WIDTH: '100%',
  STANDARD_SLIDE: '98%',
  FIRST_SLIDE: '94%',
};

export const BANNER_STYLES = {
  HEIGHT: '59px',
};

export const MAX_SLIDES = 5;
