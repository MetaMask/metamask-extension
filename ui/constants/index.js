export const ETH_DEFAULT_DECIMALS = 8;
