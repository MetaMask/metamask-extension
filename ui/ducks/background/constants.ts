import type {
  BackgroundStateProxy,
  MemStoreControllersComposedState,
} from '../../../shared/types/background';

export const initialBackgroundState: Omit<
  BackgroundStateProxy,
  keyof MemStoreControllersComposedState
> &
  Partial<{
    [ControllerName in keyof MemStoreControllersComposedState]: Partial<
      MemStoreControllersComposedState[ControllerName]
    >;
  }> = {
  isInitialized: false,
} as const;
