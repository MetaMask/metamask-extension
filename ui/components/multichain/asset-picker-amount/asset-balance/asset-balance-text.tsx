import React from 'react';
import { useSelector } from 'react-redux';
import { Text } from '../../../component-library';
import UserPreferencedCurrencyDisplay from '../../../app/user-preferenced-currency-display';
import { PRIMARY } from '../../../../helpers/constants/common';
import { Asset } from '../../../../ducks/send';
import {
  getCurrentCurrency,
  getSelectedAccountCachedBalance,
} from '../../../../selectors';
import { AssetType } from '../../../../../shared/constants/transaction';
import {
  TextColor,
  TextVariant,
} from '../../../../helpers/constants/design-system';
import CurrencyDisplay from '../../../ui/currency-display';
import { useTokenTracker } from '../../../../hooks/useTokenTracker';
import { useCurrencyDisplay } from '../../../../hooks/useCurrencyDisplay';
import { useTokenFiatAmount } from '../../../../hooks/useTokenFiatAmount';
import { getIsFiatPrimary } from '../utils';
import { useI18nContext } from '../../../../hooks/useI18nContext';
import { hexToDecimal } from '../../../../../shared/modules/conversion.utils';
import { TokenWithBalance } from '../asset-picker-modal/types';

export type AssetBalanceTextProps = {
  asset: Asset;
  balanceColor: TextColor;
  error?: string;
};

export function AssetBalanceText({
  asset,
  balanceColor,
  error,
}: AssetBalanceTextProps) {
  const t = useI18nContext();
  const secondaryCurrency = useSelector(getCurrentCurrency);

  const isFiatPrimary = useSelector(getIsFiatPrimary);

  const { tokensWithBalances }: { tokensWithBalances: TokenWithBalance[] } =
    useTokenTracker({
      tokens:
        asset.details?.address && !asset.balance
          ? [{ address: asset.details.address }]
          : [],
      address: undefined,
    });

  const balanceString =
    hexToDecimal(asset.balance) || tokensWithBalances[0]?.string;

  const balanceValue = useSelector(getSelectedAccountCachedBalance);

  const nativeTokenFiatBalance = useCurrencyDisplay(balanceValue, {
    numberOfDecimals: 2,
    currency: secondaryCurrency,
  })[1].value;

  const erc20TokenFiatBalance = useTokenFiatAmount(
    asset.details?.address,
    balanceString,
    undefined,
    // if balance is zero, conversion rate will not be available so we just assume ~0 we can't use 0 because it will set off an undefined guard
    Number(balanceString) === 0
      ? { exchangeRate: Number.MIN_VALUE }
      : undefined,
    true,
  );

  const formattedFiat =
    asset.type === AssetType.native
      ? nativeTokenFiatBalance
      : erc20TokenFiatBalance;

  const commonProps = {
    hideLabel: true,
    textProps: {
      color: balanceColor,
      variant: TextVariant.bodySm,
    },
    suffixProps: {
      color: balanceColor,
      variant: TextVariant.bodySm,
    },
  };

  const errorText = error ? `. ${t(error)}` : '';

  if (asset.type === AssetType.NFT) {
    const numberOfTokens = hexToDecimal(asset.balance || '0x0');
    return (
      <Text {...commonProps.textProps}>
        {`${numberOfTokens} ${t(
          numberOfTokens === '1' ? 'token' : 'tokens',
        )?.toLowerCase()}${errorText}`}
      </Text>
    );
  }

  if (isFiatPrimary) {
    return (
      <CurrencyDisplay
        {...commonProps}
        currency={secondaryCurrency}
        numberOfDecimals={2}
        displayValue={`${formattedFiat}${errorText}`}
      />
    );
  }

  if (asset.type === AssetType.native) {
    return (
      <>
        <UserPreferencedCurrencyDisplay
          {...commonProps}
          value={asset.balance}
          type={PRIMARY}
        />
        {errorText ? (
          <Text
            variant={TextVariant.bodySm}
            color={TextColor.errorDefault}
            data-testid="send-page-amount-error"
          >
            {errorText}
          </Text>
        ) : null}
      </>
    );
  }

  // catch-all for non-natives; they should all have addresses
  if (asset.details?.address) {
    return (
      <UserPreferencedCurrencyDisplay
        {...commonProps}
        displayValue={`${balanceString || ''}${errorText}`}
      />
    );
  }

  // this should never happen
  return null;
}
