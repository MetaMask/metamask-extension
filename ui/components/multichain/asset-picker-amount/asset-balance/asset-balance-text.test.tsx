import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { AssetType } from '../../../../../shared/constants/transaction';
import mockSendState from '../../../../../test/data/mock-send-state.json';
import configureStore from '../../../../store/store';
import { TextColor } from '../../../../helpers/constants/design-system';
import { AssetBalanceText } from './asset-balance-text';

const store = configureStore({
  ...mockSendState,
  metamask: {
    ...mockSendState.metamask,
  },
  appState: { ...mockSendState.appState, sendInputCurrencySwitched: false },
});

const mockUseTokenTracker = jest.fn();
const mockUseCurrencyDisplay = jest.fn();
const mockUseTokenFiatAmount = jest.fn();
const mockUseIsOriginalTokenSymbol = jest.fn();
const mockGetIsFiatPrimary = jest.fn();

jest.mock('../../../../hooks/useTokenTracker', () => ({
  useTokenTracker: () => mockUseTokenTracker(),
}));

jest.mock('../../../../hooks/useCurrencyDisplay', () => ({
  useCurrencyDisplay: () => mockUseCurrencyDisplay(),
}));

jest.mock('../../../../hooks/useTokenFiatAmount', () => ({
  useTokenFiatAmount: () => mockUseTokenFiatAmount(),
}));

jest.mock('../../../../hooks/useIsOriginalTokenSymbol', () => ({
  useIsOriginalTokenSymbol: () => mockUseIsOriginalTokenSymbol(),
}));

jest.mock('../utils', () => ({
  getIsFiatPrimary: () => mockGetIsFiatPrimary(),
}));

describe('AssetBalanceText', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    mockUseTokenTracker.mockReturnValue({
      tokensWithBalances: [
        { string: "doesn't matter", symbol: "doesn't matter", address: '0x01' },
      ],
    });
    mockUseCurrencyDisplay.mockReturnValue([
      'undefined',
      { value: 'fiat value', suffix: 'suffix', prefix: 'prefix-' },
    ]);
    mockUseTokenFiatAmount.mockReturnValue('Token Fiat Value');
    mockUseIsOriginalTokenSymbol.mockReturnValue(false);
    mockGetIsFiatPrimary.mockReturnValue(true);

    const asset = {
      type: AssetType.native,
      balance: '1000000',
    };
    const { asFragment } = render(
      <Provider store={store}>
        <AssetBalanceText
          asset={asset}
          balanceColor={'textColor' as TextColor}
        />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders fiat primary correctly', () => {
    mockUseTokenTracker.mockReturnValue({
      tokensWithBalances: [{ string: "doesn't matter", address: '0x01' }],
    });
    mockUseCurrencyDisplay.mockReturnValue([
      'title',
      { value: '$1.23', symbol: "doesn't matter" },
    ]);
    mockUseTokenFiatAmount.mockReturnValue("doesn't matter");
    mockUseIsOriginalTokenSymbol.mockReturnValue(true);
    mockGetIsFiatPrimary.mockReturnValue(true);

    const asset = {
      type: AssetType.token,
      details: { address: '0x01', decimals: 2 },
      balance: '100',
    };
    const { getByText } = render(
      <Provider store={store}>
        {/* Replace `any` with type */}
        {/* eslint-disable-next-line @typescript-eslint/no-explicit-any */}
        <AssetBalanceText asset={asset} balanceColor={'textColor' as any} />
      </Provider>,
    );
    expect(getByText('$1.23')).toBeInTheDocument();
  });

  it('renders native asset type correctly', () => {
    mockUseTokenTracker.mockReturnValue({
      tokensWithBalances: [{ string: '100', address: '0x01' }],
    });
    mockUseCurrencyDisplay.mockReturnValue([
      'title',
      { value: 'test_balance' },
    ]);
    mockUseTokenFiatAmount.mockReturnValue('$1.00');
    mockUseIsOriginalTokenSymbol.mockReturnValue(false);
    mockGetIsFiatPrimary.mockReturnValue(false);

    const asset = {
      type: AssetType.native,
      balance: '10000',
    };

    const { getByText } = render(
      <Provider store={store}>
        {/* Replace `any` with type */}
        {/* eslint-disable-next-line @typescript-eslint/no-explicit-any */}
        <AssetBalanceText asset={asset} balanceColor={'textColor' as any} />
      </Provider>,
    );

    expect(getByText('test_balance')).toBeInTheDocument();
  });

  it('renders error correctly', () => {
    mockUseTokenTracker.mockReturnValue({
      tokensWithBalances: [{ string: '100', address: '0x01' }],
    });
    mockUseCurrencyDisplay.mockReturnValue([
      'title',
      { value: 'test_balance' },
    ]);
    mockUseTokenFiatAmount.mockReturnValue('$1.00');
    mockUseIsOriginalTokenSymbol.mockReturnValue(false);
    mockGetIsFiatPrimary.mockReturnValue(false);

    const asset = {
      type: AssetType.native,
      balance: '10000',
    };

    const { getByText } = render(
      <Provider store={store}>
        <AssetBalanceText
          asset={asset}
          balanceColor={'textColor' as TextColor}
          error="errorText"
        />
      </Provider>,
    );

    expect(getByText('test_balance')).toBeInTheDocument();
    expect(getByText('. [errorText]')).toBeInTheDocument();
  });
});
