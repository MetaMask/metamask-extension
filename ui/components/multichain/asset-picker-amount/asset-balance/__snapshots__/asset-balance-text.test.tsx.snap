// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetBalanceText matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mm-box currency-display-component mm-box--display-flex mm-box--flex-wrap-wrap mm-box--align-items-center"
    title="undefined"
  >
    <span
      class="mm-box mm-text currency-display-component__text mm-text--body-sm mm-text--ellipsis mm-box--color-textColor"
      data-testid="account-value-and-suffix"
    >
      prefix-fiat value
    </span>
    <span
      class="mm-box mm-text currency-display-component__suffix mm-text--body-sm mm-box--margin-inline-start-1 mm-box--color-textColor"
    >
      suffix
    </span>
  </div>
</DocumentFragment>
`;
